doc/io_streams.rdoc

```diff
@@ -0,0 +1,787 @@
+- $stdout: write-only instance of \IO.
+- $stderr: read-only instance of \IO.
+- \ARGF: read-only instance of \ARGF.
+
+You can create stream objects:
+
+- \File:
+
+  - File.new: returns a new \File object.
+  - File.open: passes a new \File object to given the block.
+
+- \IO:
+
+  - IO.new: returns a new \IO object for the given integer file descriptor.
+  - IO.open: passes a new \IO object to the given block.
+  - IO.popen: returns a new \IO object that is connected to the $stdin
+    and $stdout of a newly-launched subprocess.
+  - Kernel#open: returns a new \IO object connected to a given source:
+    stream, file, or subprocess.
+
+- \StringIO:
+
+  - StringIO.new: returns a new \StringIO object.
+  - StringIO.open: passes a new \StringIO object to the given block.
+
+(You cannot create an \ARGF object, but one already exists.)
+
+=== About the Examples
+
+Many examples here use these variables:
+
+  # English text with newlines.
+  text = <<~EOT
+  First line
+  Second line
+
+  Fourth line
+  Fifth line
+  EOT
+
+  # Russian text.
+  russian = "\u{442 435 441 442}" # => "тест"
+
+  # Binary data.
+  data = "\u9990\u9991\u9992\u9993\u9994"
+
+  # Text file.
+  File.write('t.txt', text)
+
+  # File with Russian text.
+  File.write('t.rus', russian)
+
+  # File with binary data.
+  f = File.new('t.dat', 'wb:UTF-16')
+  f.write(data)
+  f.close
+
+The examples also assume that \StringIO has been required:
+
+  require 'stringio'
+
+=== Stream Modes
+
+Some stream-creating method calls specify a mode for the stream;
+the mode determines how stream may be accessed, including:
+
+- Whether the stream is to be read-only, write-only, or read-write.
+- Whether the stream treats data as text-only or binary.
+- The external and internal encodings for the stream.
+- The initial settings for stream position and line number.
+
+An \IO stream has:
+
+- A <i>read/write mode</i>, which may be read-only, write-only, or read/write;
+  see {Read/Write Mode}[rdoc-ref:io_streams.rdoc@Read-2FWrite+Mode+Specified+As+a+String].
+- A <i>data mode</i>, which may be text-only or binary;
+  see {Data Mode}[rdoc-ref:io_streams.rdoc@Data+Mode+Specified+as+a+String].
+- <i>Encodings</i>, both external and internal;
+  see {Encodings}[rdoc-ref:io_streams.rdoc@Encodings].
+- A <i>position</i>, which determines where in the stream
+  the next read or write is to occur;
+  see {Position}[rdoc-ref:io_streams.rdoc@Position].
+- A <i>line number</i>, which is a special line-oriented “position”
+  (different from the position mentioned above);
+  see {Line Number}[rdoc-ref:io_streams.rdoc@Line+Number].
+
+==== Access Mode
+
+Stream access mode may be specified as either a string or an integer.
+
+===== Read/Write Mode Specified As a \String
+
+When the mode is given as a string,
+it must begin with one of the following:
+
+- <tt>'r'</tt>: Read-only stream, positioned at the beginning:
+
+    f = File.new('t.txt', 'r')        # => #<File:t.txt>
```

Comment:
How is that better?  In each case, I've needed to illustrate by creating a file, so f.fileno is close at hand.

Reply:
It's just a bit confusing as to why you're building an `IO` object when you already have a `File` object since the `File` class inherits from the `IO` class so it can do everything an `IO` object can.

Reply:
So, remove all "io = IO.new ..."?

Reply:
Are the `IO` objects used in any way in the examples?

Reply:
No, nor the StringIO objects.  I just wanted to show how to instantiate them, and to emphasize that all are IO streams.

Reply:
Tbh I don't have strong opinions on this. On one hand, creating an `IO` from an existing `File` is odd. On the other hand, as you've said, it's nice to show how it works for each of the types (`File`, `IO`, `StringIO`).

Reply:
Not changed.